using latest buildkite plugin versions in dynamic pipeline

hnadiminti-equinix opened the pull request:
[JIRA ticket](https://packet.atlassian.net/browse/SCT-377)
Desc:
At present, we have hard coded Trivy Buildkite plugin version in Dynamic pipeline template generator. this needs to pick of latest version of the plugin everytime.

angrieralien reviewed (changes requested):
Are you able to test any of these funcs?

stephen-fox commented on util/github_util.go:
```diff
@@ -0,0 +1,20 @@
+	"fmt"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/thedevsaddam/gojsonq/v2"
```
To Stephen's point, I think both of these libraries could be replaced with libraries that Go provides:

- encoding/json - https://pkg.go.dev/encoding/json
- net/http - https://pkg.go.dev/net/http

hnadiminti-equinix replied:
The code is just picking a single field that is the latest tag from the json response.I am not sure if creating a struct and then maintaining it may pose some issues in future, in case the json response structure changes. Moreover there would be delay in unmarshalling the response to a struct.
```net/http``` is not being used because ```resty``` is much comprehensive and it provides more features like timeouts, retries etc in a more declarative fashion.

angrieralien replied:
for maintainability, i think it would be best to use the go standard libraries. Most go developers are already familiar with the http library and the json library. I'm not sure there is much benefit using these libraries over the standard libraries and we do introduce some code smells such as magic strings with `tag_name` vs using dot notation.

hnadiminti-equinix replied:
Added test cases and use the go standard libraries. Please have a look.
```
Hari.Nadiminti@EUMBQTHV267VXT dynamic-buildkite-template % go run main.go trivy
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.2:
          timeout : 5m0s
          severity: HIGH,CRITICAL
          ignore-unfixed: true
          security-checks: vuln,config%                                                                                                               
Hari.Nadiminti@EUMBQTHV267VXT dynamic-buildkite-template % go test -v ./...
?       dynamic-buildkite-template      [no test files]
?       dynamic-buildkite-template/cmd  [no test files]
?       dynamic-buildkite-template/generator    [no test files]
?       dynamic-buildkite-template/test [no test files]
=== RUN   Test_GenerateSuccess
--- PASS: Test_GenerateSuccess (0.00s)
=== RUN   Test_GenerateWrongTemplatePath
--- PASS: Test_GenerateWrongTemplatePath (0.00s)
PASS
ok      dynamic-buildkite-template/test/generator_test  (cached)
?       dynamic-buildkite-template/util [no test files]
=== RUN   Test_GetLatestTagSuccess
--- PASS: Test_GetLatestTagSuccess (0.36s)
=== RUN   Test_GetLatestTagError
--- PASS: Test_GetLatestTagError (0.15s)
PASS
ok      dynamic-buildkite-template/test/githubutil_test 1.076s
```

stephen-fox commented on util/github_util.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+func GetLatestTag(githubPAT, githubOrg, repo string) (string, error) {
+	githubAPIURL := fmt.Sprintf("https://%s:@api.github.com/repos/%s/%s", githubPAT, githubOrg, repo)
```
Why do we need to include the GitHub Personal Access Token (PAT) to make this API call?

This appears to work without authentication:

```sh
curl https://api.github.com/repos/metal-toolbox/audito-maldito/releases/latest
```

hnadiminti-equinix commented:
> Are you able to test any of these funcs?

I added the test cases and verified the functionality, Could you please check...Thanks!

prezha commented:
@hnadiminti-equinix thanks for improving your pr!

there are couple of things we can still make a bit better:
- _**unit**_ tests should be next to their "main" code, not in a separate folder/package
- not sure if `fmt.Sprintf("https://%s:@api.github.com/repos/%s/%s", githubPAT, githubOrg, repo)` will work with pat like that
  - "happy path" with private repo & pat is not tested (and would probably fail)
  - token is expected to be passed as a http request header instead - ref: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-the-latest-release
  - you should also pass additional headers in your gh api request, as noted on the ref page linked above
- using table tests instead of writing individual funcs for each test case is much more readable and maintainable
- as you pointed [above](https://github.com/equinixmetal-buildkite/dynamic-buildkite-template/pull/18#discussion_r1262400685), you only need a single key:value from the response, so you don't need to map (and maintain) all the fields in the response - just the one you're interested in, and you can use `json.Unmarshal` to unpack it

happy to jump on a call and work on these together

hnadiminti-equinix commented:
> @hnadiminti-equinix thanks for improving your pr!
> 
> there are couple of things we can still make a bit better:
> 
> * _**unit**_ tests should be next to their "main" code, not in a separate folder/package
> * not sure if `fmt.Sprintf("https://%s:@api.github.com/repos/%s/%s", githubPAT, githubOrg, repo)` will work with pat like that
>   
>   * "happy path" with private repo & pat is not tested (and would probably fail)
>   * token is expected to be passed as a http request header instead - ref: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-the-latest-release
>   * you should also pass additional headers in your gh api request, as noted on the ref page linked above
> * using table tests instead of writing individual funcs for each test case is much more readable and maintainable
> * as you pointed [above](https://github.com/equinixmetal-buildkite/dynamic-buildkite-template/pull/18#discussion_r1262400685), you only need a single key:value from the response, so you don't need to map (and maintain) all the fields in the response - just the one you're interested in, and you can use `json.Unmarshal` to unpack it
> * in general, when calling an api, you should use sensible timeout, to avoid call that hangs and block code/tests
> 
> happy to jump on a call and work on these together

Hi @prezha I have addressed all of the review comments; Could you please check and let me know.
```
Hari.Nadiminti@EUMBQTHV267VXT dynamic-buildkite-template % go test -v ./...    
?       dynamic-buildkite-template      [no test files]
?       dynamic-buildkite-template/cmd  [no test files]
=== RUN   Test_GenerateTrivyStep
=== RUN   Test_GenerateTrivyStep/success
=== RUN   Test_GenerateTrivyStep/wrong_template_path
--- PASS: Test_GenerateTrivyStep (0.00s)
    --- PASS: Test_GenerateTrivyStep/success (0.00s)
    --- PASS: Test_GenerateTrivyStep/wrong_template_path (0.00s)
PASS
ok      dynamic-buildkite-template/generator    (cached)
=== RUN   Test_GetLatestTag
=== RUN   Test_GetLatestTag/PAT_provided
=== RUN   Test_GetLatestTag/PAT_absent
--- PASS: Test_GetLatestTag (0.48s)
    --- PASS: Test_GetLatestTag/PAT_provided (0.35s)
    --- PASS: Test_GetLatestTag/PAT_absent (0.13s)
PASS
ok      dynamic-buildkite-template/util 0.830s
Hari.Nadiminti@EUMBQTHV267VXT dynamic-buildkite-template % go run main.go trivy
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.2:
          timeout : 5m0s
          severity: HIGH,CRITICAL
          ignore-unfixed: true
          security-checks: vuln,config
          
 ```

prezha commented on generator/common-pipeline_test.go:
```diff
@@ -0,0 +1,60 @@
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
```
can we avoid using this additional package?
it looks like we can replace it with simple if/else constructs + eg, [t.Fatalf()](https://pkg.go.dev/testing#T.Fatalf)  instead of using require.ErrorContains() and require.Equal()?

hnadiminti-equinix replied:
Addressing the code modifications

prezha commented on util/githubutil_test.go:
```diff
@@ -0,0 +1,62 @@
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/require"
```
same comment as above

hnadiminti-equinix replied:
Addressing the code modifications

prezha commented on generator/common-pipeline_test.go:
```diff
@@ -0,0 +1,60 @@
+	expVal       string
+}
+
+func Test_GenerateTrivyStep(t *testing.T) {
```
no need to use `_` in func name

hnadiminti-equinix replied:
Addressing the code modifications

prezha commented on util/githubutil_test.go:
```diff
@@ -0,0 +1,62 @@
+	Repo string
+}
+
+func Test_GetLatestTag(t *testing.T) {
```
same comment as above re: func naming

hnadiminti-equinix replied:
Addressing the code modifications

prezha commented:
thank you @hnadiminti-equinix for further improving your pr!

i think we're getting there - please see last few comments and also, for consistency:
- util/github_util.go should be util/github.go
- util/githubutil_test.go should be util/github_test.go

hnadiminti-equinix commented:
> thank you @hnadiminti-equinix for further improving your pr!
> 
> i think we're getting there - please see last few comments and also, for consistency:
> 
> * util/github_util.go should be util/github.go
> * util/githubutil_test.go should be util/github_test.go

Done, Thanks @prezha Can you please check and let me know if I need to do any further changes. Ta!!

prezha reviewed (approved):
thank you @hnadiminti-equinix for your patience and for improving the pr!

if other reviewers are satisfied with the changes you've made as well (make sure to "resolve" the comments you've addressed), i'm happy to merge it
